api/resource/resource.go

```diff
@@ -290,9 +290,18 @@ func (r *Resource) getCsvAnnotation(name string) []string {
 // PrefixesSuffixesEquals is conceptually doing the same task
 // as OutermostPrefixSuffix but performs a deeper comparison
 // of the suffix and prefix slices.
-func (r *Resource) PrefixesSuffixesEquals(o ResCtx) bool {
-	return utils.SameEndingSubSlice(r.GetNamePrefixes(), o.GetNamePrefixes()) &&
-		utils.SameEndingSubSlice(r.GetNameSuffixes(), o.GetNameSuffixes())
+func (r *Resource) PrefixesSuffixesEquals(o ResCtx, allowEmpty bool) bool {
+	if allowEmpty {
```

Comment:
Do we really need this condition here?

From what I've seen in this PR, you'll run the function twice, one with the flag set to `true`, another one with the flag set to `false`. Is there a specific reason why applying the OR would be undesirable?

Reply:
Are you asking, could we just go with the following function?

```go
func (r *Resource) PrefixesSuffixesEquals(o ResCtx) bool {                                                 
        eitherPrefixEmpty := len(r.GetNamePrefixes()) == 0 || len(o.GetNamePrefixes()) == 0                
        eitherSuffixEmpty := len(r.GetNameSuffixes()) == 0 || len(o.GetNameSuffixes()) == 0                
                                                                                                           
        return (eitherPrefixEmpty || utils.SameEndingSubSlice(r.GetNamePrefixes(), o.GetNamePrefixes())) &&
                (eitherSuffixEmpty || utils.SameEndingSubSlice(r.GetNameSuffixes(), o.GetNameSuffixes()))  
}                                                                                                          
```

This causes regressions with `TestIssue3489Simplified`, `TestIssue3489` and `TestNameReferenceDeploymentIssue3489`, all with "found multiple possible referrals" errors.

I've tried to explain this problem/the motivation behind the `allowEmpty` option here https://github.com/kubernetes-sigs/kustomize/issues/5047#issuecomment-1627542060.

Reply:
Thanks for the review and apologies I didn't get around to writing a description beforehand! 🙏 
My memory is slightly hazy, as it's been a while since I worked on this, but I think your description is accurate 🙂 
Expanding on a couple of points:

>  if we simply replace that with an or

i.e. allow either (inclusive) a prefix or suffix to match

>  it expects both prefixes and suffixes to be the same

And if a prefix/suffix is _missing_ it treats them as not the same (which wasn't the case before https://github.com/kubernetes-sigs/kustomize/pull/3529, where an empty prefix/suffix was considered to have the same SameEndingSubarray as a _one_ element prefix/suffix).

I also think https://github.com/kubernetes-sigs/kustomize/issues/5047#issuecomment-1627536967 and https://github.com/kubernetes-sigs/kustomize/issues/5047#issuecomment-1627542060 are the  most useful/accurate comments I've left.

Reply:
Just a small suggestion: I think it's worth adding a comment with a brief context of why the condition is needed, for future reference.

Reply:
>  The second pass is only going to happen if/when the resulting list of candidates is bigger than one element

Yep, exactly that!

I totally agree about adding a comment explaining this, I'll try and write something as soon as possible — I only held off before as I wasn't sure if this was even the correct/best approach.

Thanks again for the review! 🙏

Reply:
@stormqueen1990 I've added a comment and update the PR description.
Let me know if you want any changes/think more the of PR description should live in the code as a comment 🙂

Reply:
@jonathanlking this looks good, thanks for adding!